lib/apiservers/engine/backends/container.go

```diff
@@ -345,137 +334,288 @@ func (c *ContainerBackend) ContainerExecResize(eid string, height, width int) er
+		waitState := map[string]bool{constants.TaskCreatedState: true, constants.TaskUnknownState: true}
+		_, err := c.taskStateWaitHelper(taskOp, id, eid, name, targetState, waitState)
+		if err != nil {
+			op.Errorf("Wait for exec start on %s.%s: %s", eid, id, err)
+			startResult <- err
+		}
+
+		close(startResult)
+	}()
+
+	// wait for either wait to succeed, fail, or for stream connection to error out
+	select {
+	case err := <-startResult:
+		if err != nil {
+			op.Errorf("Task wait returned error: %s", err)
+			// we can't return a proper error as we close the streams as soon as AttachStreams returns so we mimic Docker and write to stdout directly
+			// https://github.com/docker/docker/blob/a039ca9affe5fa40c4e029d7aae399b26d433fe9/api/server/router/container/exec.go#L114
+			if stdout != nil {
+				stdout.Write([]byte(err.Error() + "\r\n"))
 			}
 
+			// This will cause attachHelper to exit
+			cancel()
 			return err
 		}
+
+	case err := <-attachResult:
+		// we pass a nil actor as we don't want detach events dispatched at this point
+		detach, aerr := processAttachError(op, nil, err)
+		if aerr != nil {
+			stdin.Close()
+			cancel()
+			return aerr
+		}
+
+		if detach {
+			// put it back on the queue to process after attach event is reported
+			op.Debugf("Requeuing early detach")
+			attachResult = make(chan error, 1)
+			attachResult <- engerr.DetachError{}
+		}
+	}
+
+	op.Infof("Exec %s in %s launched successfully", id, eid)
+
+	actor := CreateContainerEventActorWithAttributes(vc, map[string]string{})
+
+	// exec_start event
+	event := "exec_start: " + ec.ProcessConfig.ExecPath + " " + strings.Join(ec.ProcessConfig.ExecArgs[1:], " ")
+	EventService().Log(event, eventtypes.ContainerEventType, actor)
+
+	// no need to attach for detached case
+	if !attach {
+		op.Debugf("Detached mode. Returning early.")
 		return nil
 	}
 
-	// configure custom exec back off configure
-	backoffConf := retry.NewBackoffConfig()
-	backoffConf.MaxInterval = 2 * time.Second
-	backoffConf.InitialInterval = 500 * time.Millisecond
+	EventService().Log(containerAttachEvent, eventtypes.ContainerEventType, actor)
+	// we don't use autoclose so ensure the transport is shut down on exit
+	defer stdin.Close()
 
-	if err := retry.DoWithConfig(operation, engerr.IsConflictError, backoffConf); err != nil {
-		op.Errorf("Failed to start Exec task for container(%s) due to error (%s)", id, err)
+	// wait for attach streams to complete if attaching
+	detach, aerr := processAttachError(op, &actor, <-attachResult)
+	if aerr != nil || detach {
+		op.Debugf("Skipping task wait, err: %s, detach: %t", aerr, detach)
+		// if detached then no expectation the process will exit, so don't wait on it
+		return aerr
+	}
+
+	op.Debugf("Waiting for completion: task(%s), container(%s)", eid, name)
+
+	targetState := map[string]bool{constants.TaskStoppedState: true}
+	waitState := map[string]bool{constants.TaskRunningState: true, constants.TaskUnknownState: true}
```

Comment:
Should include `constants.TaskCreatedState` - it's possible to get here via the attach success path while still waiting for the host sync to get the `TaskRunningState` update.